ddtrace/internal/flare.py

```diff
@@ -0,0 +1,158 @@
+import datetime
+import json
+import os
+import pkg_resources
+import platform
+import sys
+
+import ddtrace
+from ddtrace.utils import formats
+
+from .logger import get_logger
+
+
+logger = get_logger(__name__)
+
+
+def ping_agent(api=None, hostname=None, port=None, uds_path=None):
+    # Attempt to query the agent, returns an api.Response
+    # or one of the following exceptions: httplib.HTTPException, OSError, IOError
+
+    if not api:
+        api = ddtrace.api.API(hostname=hostname, port=port, uds_path=uds_path,)
+
+    # We can't use api.send_traces([]) since it'll shortcut
+    # if traces is falsy.
+    p = ddtrace.payload.Payload(encoder=api._encoder)
+
+    # We can't use payload.add_trace([]) for the same reason
+    # as api.send_trace([]).
+    encoded = p.encoder.encode_trace([])
+    p.traces.append(encoded)
+    p.size += len(encoded)
+
+    resp = api._flush(p)
+
+    return resp
+
+
+def in_venv():
+    # Works with both venv and virtualenv
+    # https://stackoverflow.com/a/42580137
+    if "VIRTUAL_ENV" in os.environ:
+        return True
+
+    return hasattr(sys, "real_prefix") or (hasattr(sys, "base_prefix") and sys.base_prefix != sys.prefix)
+
+
+def tags_to_str(tags):
+    # Turn a dict of tags to a string "k1:v1,k2:v2,..."
+    return ",".join(["%s:%s" % (k, v) for k, v in tags.items()])
+
+
+def flare(tracer):
+    """Collect system and library information
+    """
+
+    # The tracer doesn't actually maintain a hostname/port, instead it stores
+    # it on the possibly None writer which actually stores it on an API object.
+    # Note that the tracer DOES have hostname and port attributes that it
+    # sets to the defaults and ignores.
+    if tracer.writer:
+        hostname = tracer.writer.api.hostname
+        port = tracer.writer.api.port
+        uds_path = tracer.writer.api.uds_path
+
+        # If all specified, uds_path will take precedence
+        if uds_path:
+            agent_url = "uds://%s" % uds_path
+        else:
+            agent_url = "http://%s:%s" % (hostname, port)
+    else:
+        # Else if we can't infer anything from the tracer, rely on the defaults.
+        hostname = ddtrace.Tracer.DEFAULT_HOSTNAME
+        port = ddtrace.Tracer.DEFAULT_PORT
+        agent_url = "http://%s:%s" % (hostname, port)
```

Comment:
Does it plays well with `DD_TRACE_AGENT_URL` which seems to be [supported](https://docs.datadoghq.com/tracing/setup/python/#instrumentation/) in python?

Reply:
Yes, since it gets [parsed into the hostname and port attributes](https://github.com/DataDog/dd-trace-py/blob/0d7e7cb38216acb0c8b29f0ae1318d25bc160123/ddtrace/tracer.py#L100-L114).